ably/transport/defaults.py

```diff
@@ -46,13 +42,53 @@ def get_port(options):
                 return Defaults.port
 
     @staticmethod
+    def get_rest_hosts(options):
+        """
+        Return the list of hosts as they should be tried. First comes the main
+        host. Then the fallback hosts in random order.
+        The returned list will have a length of up to http_max_retry_count.
+        """
+        # Defaults
+        host = options.rest_host
+        if host is None:
+            host = Defaults.rest_host
+
+        environment = options.environment
+        if environment is None:
+            environment = Defaults.environment
+
+        http_max_retry_count = options.http_max_retry_count
+        if http_max_retry_count is None:
+            http_max_retry_count = Defaults.http_max_retry_count
+
+        # Prepend environment
+        if environment != 'production':
+            host = '%s-%s' % (environment, host)
+
+        # Fallback hosts
+        fallback_hosts = options.fallback_hosts
+        if fallback_hosts is None:
+            if host == Defaults.rest_host or options.fallback_hosts_use_default:
+                fallback_hosts = Defaults.fallback_hosts
+            else:
+                fallback_hosts = []
+
+        # Shuffle
+        fallback_hosts = list(fallback_hosts)
+        random.shuffle(fallback_hosts)
+
+        # First main host
+        hosts = [host] + fallback_hosts
+        hosts = hosts[:http_max_retry_count]
+        return hosts
```

Comment:
Should we not memoize / cache this to avoid having to determine this for every request.  As the client options are immutable, there is not need to recalculate this each time as far as I am aware.